add_git_tag error: No value found for build_number

<!-- Thanks for helping fastlane! Before you submit your issue, please make sure you followed our checklist and check the appropriate boxes by putting an x in the [ ]: [x] -->

### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
I'm trying to deploy my Android app via Fastlane and add a git tag after deploy to Firebase. I've added a custom tag but Fastlane keeps complaining about: 'No value found for 'build_number''. Everything else works fine.

##### Command executed
```
private_lane :add_tag_to_git do
    get_android_version(apk: lane_context[SharedValues::GRADLE_APK_OUTPUT_PATH])

    version = "v#{lane_context[SharedValues::GET_ANDROID_VERSION_NAME]}_#{lane_context[SharedValues::GET_ANDROID_VERSION_CODE]}"
    add_git_tag(
      tag: "Firebase/#{current_env}/#{version}"
    )
  end
```

##### Complete output when running fastlane, including the stack trace and command used

> Run bundle exec fastlane deploy_integration
  bundle exec fastlane deploy_integration
  shell: /bin/bash -e {0}
  env:
    KEYSTORE_PASS: ***
    KEYSTORE_ALIAS: ***
    KEYSTORE_ALIAS_PASS: ***
    RELEASE_SLACK_URL: ***
    FIREBASE_TOKEN: ***
    FIREBASE_APP_ID_INTEGRATION: ***
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:29: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:30: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
/usr/lib/ruby/vendor_ruby/rubygems/defaults/operating_system.rb:10: warning: constant Gem::ConfigMap is deprecated
+---------------------------------------------+---------+---------------------------------+
|                                      Used plugins                                       |
+---------------------------------------------+---------+---------------------------------+
| Plugin                                      | Version | Action                          |
+---------------------------------------------+---------+---------------------------------+
| fastlane-plugin-firebase_app_distribution   | 0.2.4   | firebase_app_distribution       |
|                                             |         | firebase_app_distribution_login |
| fastlane-plugin-commit_android_version_bump | 0.2.0   | commit_android_version_bump     |
| fastlane-plugin-versioning_android          | 0.1.0   | android_get_version_name        |
|                                             |         | android_get_version_code        |
|                                             |         | android_set_version_code        |
|                                             |         | android_set_version_name        |
| fastlane-plugin-get_android_version         | 0.3.5   | get_android_version             |
+---------------------------------------------+---------+---------------------------------+

[08:32:01]: Sending anonymous analytics information
[08:32:01]: Learn more at https://docs.fastlane.tools/#metrics
[08:32:01]: No personal or sensitive data is sent.
[08:32:01]: You can disable this by adding `opt_out_usage` at the top of your Fastfile
[08:32:01]: ------------------------------
[08:32:01]: --- Step: default_platform ---
[08:32:01]: ------------------------------
[08:32:01]: Driving the lane 'android deploy_integration' 🚀
[08:32:01]: -------------------
[08:32:01]: --- Step: is_ci ---
[08:32:01]: -------------------
> [08:32:49]: -------------------------------
[08:32:49]: --- Step: ensure_git_branch ---
[08:32:49]: -------------------------------
[08:32:49]: Git branch matches `integration`, all good! 💪
[08:32:49]: -------------------
[08:32:49]: --- Step: clean ---
[08:32:49]: -------------------
[08:32:49]: $ /home/runner/work/tessa-android/tessa-android/gradlew clean -p .
[08:32:50]: ▸ > Configure project :app
[08:32:50]: ▸ WARNING: API 'BaseVariant.getApplicationIdTextResource' is obsolete and has been replaced with 'VariantProperties.applicationId'.
[08:32:50]: ▸ It will be removed in version 5.0 of the Android Gradle plugin.
[08:32:50]: ▸ For more information, see TBD.
[08:32:50]: ▸ To determine what is calling BaseVariant.getApplicationIdTextResource, use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
[08:32:50]: ▸ app: 'annotationProcessor' dependencies won't be recognized as kapt annotation processors. Please change the configuration name to 'kapt' for these artifacts: 'com.github.bumptech.glide:compiler:4.11.0'.
[08:32:50]: ▸ > Task :clean UP-TO-DATE
[08:32:50]: ▸ > Task :app:clean
[08:32:50]: ▸ BUILD SUCCESSFUL in 1s
[08:32:50]: ▸ 2 actionable tasks: 1 executed, 1 up-to-date
[08:32:50]: ----------------------------------------
[08:32:50]: --- Step: assembleintegrationRelease ---
[08:32:50]: ----------------------------------------
[08:32:50]: $ /home/runner/work/tessa-android/tessa-android/gradlew assembleintegrationRelease -p . -Pandroid.injected.signing.store.file=Tessa.jks -Pandroid.injected.signing.store.password=*** -Pandroid.injected.signing.key.alias=*** -Pandroid.injected.signing.key.password=***
[08:32:52]: ▸ > Configure project :app
[08:32:52]: ▸ WARNING: API 'BaseVariant.getApplicationIdTextResource' is obsolete and has been replaced with 'VariantProperties.applicationId'.
[08:32:52]: ▸ It will be removed in version 5.0 of the Android Gradle plugin.
[08:32:52]: ▸ For more information, see TBD.
[08:32:52]: ▸ To determine what is calling BaseVariant.getApplicationIdTextResource, use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
[08:32:52]: ▸ app: 'annotationProcessor' dependencies won't be recognized as kapt annotation processors. Please change the configuration name to 'kapt' for these artifacts: 'com.github.bumptech.glide:compiler:4.11.0'.
[08:33:17]: ▸ > Task :app:preBuild UP-TO-DATE
[08:33:17]: ▸ > Task :app:preIntegrationReleaseBuild UP-TO-DATE
[08:33:17]: ▸ > Task :app:compileIntegrationReleaseAidl NO-SOURCE
[08:33:17]: ▸ > Task :app:compileIntegrationReleaseRenderscript NO-SOURCE
[08:33:18]: ▸ > Task :app:dataBindingMergeDependencyArtifactsIntegrationRelease
[08:33:18]: ▸ > Task :app:dataBindingMergeGenClassesIntegrationRelease
[08:33:18]: ▸ > Task :app:generateIntegrationReleaseResValues
[08:33:18]: ▸ > Task :app:generateIntegrationReleaseResources
[08:33:18]: ▸ > Task :app:injectCrashlyticsMappingFileIdIntegrationRelease
[08:33:18]: ▸ > Task :app:processIntegrationReleaseGoogleServices
[08:33:21]: ▸ > Task :app:dataBindingTriggerIntegrationRelease
[08:33:21]: ▸ > Task :app:generateIntegrationReleaseBuildConfig
[08:33:21]: ▸ > Task :app:checkApolloVersions
[08:33:21]: ▸ > Task :app:mergeIntegrationReleaseResources
[08:33:21]: ▸ > Task :app:generateIntegrationReleaseServiceApolloSources
[08:33:22]: ▸ > Task :app:dataBindingGenBaseClassesIntegrationRelease
[08:33:22]: ▸ > Task :app:writeIntegrationReleaseApplicationId
[08:33:23]: ▸ > Task :app:generateSafeArgsIntegrationRelease
[08:33:23]: ▸ > Task :app:checkIntegrationReleaseAarMetadata
[08:33:23]: ▸ > Task :app:createIntegrationReleaseCompatibleScreenManifests
[08:33:23]: ▸ > Task :app:extractDeepLinksIntegrationRelease
[08:33:23]: ▸ > Task :app:processIntegrationReleaseMainManifest
[08:33:23]: ▸ > Task :app:processIntegrationReleaseManifest
[08:33:23]: ▸ > Task :app:processIntegrationReleaseManifestForPackage
[08:33:26]: ▸ > Task :app:processIntegrationReleaseResources
[08:33:54]: ▸ > Task :app:kaptGenerateStubsIntegrationReleaseKotlin
[08:33:57]: ▸ > Task :app:javaPreCompileIntegrationRelease
[08:33:58]: ▸ > Task :app:mergeIntegrationReleaseNativeDebugMetadata NO-SOURCE
[08:33:58]: ▸ > Task :app:mergeIntegrationReleaseShaders
[08:33:58]: ▸ > Task :app:compileIntegrationReleaseShaders NO-SOURCE
[08:33:58]: ▸ > Task :app:generateIntegrationReleaseAssets UP-TO-DATE
[08:33:58]: ▸ > Task :app:mergeIntegrationReleaseAssets
[08:33:58]: ▸ > Task :app:compressIntegrationReleaseAssets
[08:34:01]: ▸ > Task :app:checkIntegrationReleaseDuplicateClasses
[08:34:01]: ▸ > Task :app:processIntegrationReleaseJavaRes NO-SOURCE
[08:34:01]: ▸ > Task :app:kaptIntegrationReleaseKotlin
[08:34:01]: ▸ /home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/entities/Asset.java:11: warning: companyId column references a foreign key but it is not part of an index. This may trigger full table scans whenever parent table is modified so you are highly advised to create an index that covers this column.
[08:34:01]: ▸ public final class Asset implements java.io.Serializable {
[08:34:01]: ▸ ^/home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/entities/Asset.java:11: warning: locationId column references a foreign key but it is not part of an index. This may trigger full table scans whenever parent table is modified so you are highly advised to create an index that covers this column.
[08:34:01]: ▸ public final class Asset implements java.io.Serializable {
[08:34:01]: ▸ ^/home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/entities/CompanyLocation.java:20: warning: Primary key constraint on addressId is ignored when being merged into nl.sping.tessa.data.model.room.entities.CompanyLocation
[08:34:01]: ▸ private final nl.sping.tessa.data.model.room.entities.Address address = null;
[08:34:01]: ▸ ^/home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/entities/CompanyLocation.java:7: warning: companyId column references a foreign key but it is not part of an index. This may trigger full table scans whenever parent table is modified so you are highly advised to create an index that covers this column.
[08:34:01]: ▸ public final class CompanyLocation {
[08:34:01]: ▸ ^/home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/entities/ProductSubGroup.java:7: warning: productGroupId column references a foreign key but it is not part of an index. This may trigger full table scans whenever parent table is modified so you are highly advised to create an index that covers this column.
[08:34:01]: ▸ public final class ProductSubGroup {
[08:34:01]: ▸ ^/home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/entities/Product.java:7: warning: productSubGroupId column references a foreign key but it is not part of an index. This may trigger full table scans whenever parent table is modified so you are highly advised to create an index that covers this column.
[08:34:01]: ▸ public final class Product {
[08:34:01]: ▸ ^/home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/entities/PropertyValue.java:7: warning: productId column references a foreign key but it is not part of an index. This may trigger full table scans whenever parent table is modified so you are highly advised to create an index that covers this column.
[08:34:01]: ▸ public final class PropertyValue {
[08:34:01]: ▸ ^/home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/entities/PropertyValue.java:7: warning: productPropertyId column references a foreign key but it is not part of an index. This may trigger full table scans whenever parent table is modified so you are highly advised to create an index that covers this column.
[08:34:01]: ▸ public final class PropertyValue {
[08:34:01]: ▸ ^/home/runner/work/tessa-android/tessa-android/app/build/tmp/kapt3/stubs/integrationRelease/nl/sping/tessa/data/model/room/AppDatabase.java:8: warning: Schema export directory is not provided to the annotation processor so we cannot export the schema. You can either provide `room.schemaLocation` annotation processor argument OR set exportSchema to false.
[08:34:01]: ▸ public abstract class AppDatabase extends androidx.room.RoomDatabase {
[08:34:02]: ▸ ^ANTLR Tool version 4.5.3 used for code generation does not match the current runtime version 4.7.1ANTLR Runtime version 4.5.3 used for parser compilation does not match the current runtime version 4.7.1ANTLR Tool version 4.5.3 used for code generation does not match the current runtime version 4.7.1ANTLR Runtime version 4.5.3 used for parser compilation does not match the current runtime version 4.7.1
[08:34:02]: ▸ > Task :app:collectIntegrationReleaseDependencies
[08:34:02]: ▸ > Task :app:sdkIntegrationReleaseDependencyData
[08:34:02]: ▸ > Task :app:mergeIntegrationReleaseJniLibFolders
[08:34:03]: ▸ > Task :app:kaptIntegrationReleaseKotlin
[08:34:03]: ▸ [WARN] Incremental annotation processing requested, but support is disabled because the following processors are not incremental: androidx.room.RoomProcessor (DYNAMIC).
[08:34:05]: ▸ > Task :app:mergeIntegrationReleaseNativeLibs
[08:34:23]: ▸ > Task :app:compileIntegrationReleaseKotlin
[08:34:23]: ▸ w: /home/runner/work/tessa-android/tessa-android/app/src/main/java/nl/sping/tessa/ui/asset/ScanForAssetFragment.kt: (236, 42): Candidate resolution will be changed soon, please use fully qualified name to invoke the following closer candidate explicitly '
[08:34:23]: ▸ public open fun observe(p0: LifecycleOwner, p1: Observer<in Int!>): Unit defined in androidx.lifecycle.LiveData'
[08:34:23]: ▸ w: /home/runner/work/tessa-android/tessa-android/app/src/main/java/nl/sping/tessa/viewModel/AssetInspectionViewModel.kt: (50, 26): Parameter 'attachment' is never used
[08:34:23]: ▸ w: /home/runner/work/tessa-android/tessa-android/app/src/main/java/nl/sping/tessa/viewModel/AssetInspectionViewModel.kt: (70, 26): Parameter 'uri' is never used
[08:34:31]: ▸ > Task :app:compileIntegrationReleaseJavaWithJavac
[08:34:31]: ▸ > Task :app:compileIntegrationReleaseSources
[08:35:01]: ▸ > Task :app:lintVitalIntegrationRelease
[08:35:14]: ▸ > Task :app:transformClassesWithFirebasePerformancePluginForIntegrationRelease
[08:35:44]: ▸ > Task :app:dexBuilderIntegrationRelease
[08:36:13]: ▸ > Task :app:mergeExtDexIntegrationRelease
[08:36:14]: ▸ > Task :app:mergeIntegrationReleaseJavaResource
[08:36:15]: ▸ > Task :app:stripIntegrationReleaseDebugSymbols
[08:36:15]: ▸ > Task :app:validateSigningIntegrationRelease
[08:36:22]: ▸ > Task :app:mergeDexIntegrationRelease
[08:36:24]: ▸ > Task :app:packageIntegrationRelease
[08:36:24]: ▸ > Task :app:assembleIntegrationRelease
[08:36:24]: ▸ Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
[08:36:24]: ▸ Use '--warning-mode all' to show the individual deprecation warnings.
[08:36:24]: ▸ See https://docs.gradle.org/6.5/userguide/command_line_interface.html#sec:command_line_warnings
[08:36:24]: ▸ BUILD SUCCESSFUL in 3m 33s
[08:36:24]: ▸ 42 actionable tasks: 42 executed
[08:36:24]: -------------------------------------------------------
[08:36:24]: --- Step: Switch to android upload_to_firebase lane ---
[08:36:24]: -------------------------------------------------------
[08:36:24]: Cruising over to lane 'android upload_to_firebase' 🚖
[08:36:24]: ---------------------------------------
[08:36:24]: --- Step: firebase_app_distribution ---
[08:36:24]: ---------------------------------------
[08:36:24]: Authenticating with FIREBASE_TOKEN environment variable
[08:36:24]: 🔐 Authenticated successfully.
[08:36:25]: ⌛ Uploading the APK.
[08:36:36]: ✅ Uploaded the APK.
[08:36:36]: ✅ No release notes passed in. Skipping this step.
[08:36:37]: ✅ Added testers/groups.
[08:36:37]: 🎉 App Distribution upload finished successfully.
[08:36:37]: Cruising back to lane 'android firebase_deploy' 🚘
[08:36:37]: -------------------------
[08:36:37]: --- Step: add_git_tag ---
[08:36:37]: -------------------------
+-------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
|                                                                         Lane Context                                                                          |
+-------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
| DEFAULT_PLATFORM                    | android                                                                                                                 |
| PLATFORM_NAME                       | android                                                                                                                 |
| LANE_NAME                           | android deploy_integration                                                                                              |
| GRADLE_BUILD_TYPE                   | integrationRelease                                                                                                      |
| GRADLE_ALL_APK_OUTPUT_PATHS         | ["/home/runner/work/tessa-android/tessa-android/app/build/outputs/apk/integration/release/app-integration-release.apk"] |
| GRADLE_ALL_AAB_OUTPUT_PATHS         | []                                                                                                                      |
| GRADLE_ALL_OUTPUT_JSON_OUTPUT_PATHS | []                                                                                                                      |
| GRADLE_ALL_MAPPING_TXT_OUTPUT_PATHS | []                                                                                                                      |
| GRADLE_APK_OUTPUT_PATH              | /home/runner/work/tessa-android/tessa-android/app/build/outputs/apk/integration/release/app-integration-release.apk     |
+-------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
[08:36:37]: No value found for 'build_number'


[!] No value found for 'build_number'
+------+-------------------------------------------+-------------+
|                        fastlane summary                        |
+------+-------------------------------------------+-------------+
| Step | Action                                    | Time (in s) |
+------+-------------------------------------------+-------------+
| 1    | default_platform                          | 0           |
| 2    | is_ci                                     | 0           |
| 3    | clean                                     | 45          |
| 4    | fastlane install_plugins                  | 1           |
| 5    | Switch to android firebase_deploy lane    | 0           |
| 6    | is_ci                                     | 0           |
| 7    | ensure_git_branch                         | 0           |
| 8    | clean                                     | 1           |
| 9    | assembleintegrationRelease                | 214         |
| 10   | Switch to android upload_to_firebase lane | 0           |
| 11   | firebase_app_distribution                 | 12          |
| 💥   | add_git_tag                               | 0           |
+------+-------------------------------------------+-------------+

[08:36:37]: fastlane finished with errors
Error: Process completed with exit code 1.

--- a/fastlane/lib/fastlane/actions/add_git_tag.rb
+++ b/fastlane/lib/fastlane/actions/add_git_tag.rb
@@ -6,7 +6,13 @@ module Fastlane
         # lane name in lane_context could be nil because you can just call $fastlane add_git_tag which has no context
         lane_name = Actions.lane_context[Actions::SharedValues::LANE_NAME].to_s.delete(' ') # no spaces allowed
 
-        tag = options[:tag] || "#{options[:grouping]}/#{lane_name}/#{options[:prefix]}#{options[:build_number]}#{options[:postfix]}"
+        if options[:tag]
+          tag = options[:tag]
+        elsif options[:build_number]
+          tag = "#{options[:grouping]}/#{lane_name}/#{options[:prefix]}#{options[:build_number]}#{options[:postfix]}"
+        else
+          UI.user_error!("No value found for 'tag' or 'build_number'. At least one of them must be provided. Note that if you do specify a tag, all other arguments are ignored.")
+        end
         message = options[:message] || "#{tag} (fastlane)"
 
         cmd = ['git tag']
@@ -64,7 +70,8 @@ module Fastlane
                                        description: "The build number. Defaults to the result of increment_build_number if you\'re using it",
                                        default_value: Actions.lane_context[Actions::SharedValues::BUILD_NUMBER],
                                        default_value_dynamic: true,
-                                       is_string: false),
+                                       is_string: false,
+                                       optional: true),
           FastlaneCore::ConfigItem.new(key: :message,
                                        env_name: "FL_GIT_TAG_MESSAGE",
                                        description: "The tag message. Defaults to the tag's name",

--- a/fastlane/spec/actions_specs/add_git_tag_spec.rb
+++ b/fastlane/spec/actions_specs/add_git_tag_spec.rb
@@ -76,6 +76,16 @@ describe Fastlane do
         expect(result).to eq("git tag -am #{message.shellescape} #{tag.shellescape}")
       end
 
+      it "raises error if no tag or build_number are provided" do
+        Fastlane::Actions.lane_context[Fastlane::Actions::SharedValues::BUILD_NUMBER] = nil
+
+        expect do
+          Fastlane::FastFile.new.parse("lane :test do
+            add_git_tag ({})
+          end").runner.execute(:test)
+        end.to raise_error(/No value found for 'tag' or 'build_number'. At least one of them must be provided. Note that if you do specify a tag, all other arguments are ignored./)
+      end
+
       it "specified tag overrides generate tag" do
         tag = '2.0.0'
 

